### ISSUE TITLE
Support/Document way to extend Logstash's classpath

### ISSUE THREAD

**Issue by n0othing** [elastic/logstash#6874]
Certain plugins require extending the classpath (kafka input/output for custom serializers), and we leave it up to the end user to figure out how to approach it.

For example, _today_ you can add your library files to the `CLASSPATH` environment variable, or drop them in `vendor/jruby/lib`

It'd be great to have a tried n' true (and documented) way to do this. Maybe an option in the `jvm.options` file?

**Comment by andrewvc** [comment]
@n0othing good point! IMHO the `CLASSPATH` env variables seems like the way to go.

@jordansissel WDYT? Maybe we just need to document this?

**Comment by anelson-vidscale** [comment]
The problem is sharing and ordering the classpath among the various interested parties.
For me, I wanted both the Logstash classpath *and* the classpath required for the serializers.
I did not find any way to share.  I was able to hack it up to use either the LogStash classpath or my serializers CLASSPATH, but I found no way to join them. 
I had to give up and put softlinks to my library files in vendor/jruby/lib -- an approach that feels hacky and un-blessed by elastic.co.

**Comment by jordansissel** [comment]
For only plugins that need it, how about this:

1) We provide a custom class loader to the plugin
2) That only has specific libraries available
3) And is extendable by the operator with a setting provided by the plugin

Thoughts?

I am not in favor of adding a global way to inject code into Logstash, but having something specific per plugin (not all plugins need this, perhaps only Kafka?).

On the kafka note, can we talk about the specific use case?

**Comment by anelson-vidscale** [comment]
I discussed my specific use case with @n0othing
<https://github.com/n0othing> in an elastic.co support ticket.
If you'd like, I can repeat the conversation here.
I'm fine with a plugin custom class loader.
I'm fine with any solution that is "blessed" by elastic.

Okay, some quick detail.
I want a Kafka plugin that includes the Kafka API as
not-included-dependency because I wish to use the mapr.com Kafka
implementation and that plugin should allow me to say where the jar(s) are
coming from (e..g via a CLASSPATH like syntax).

**Comment by jordansissel** [comment]
> because I wish to use the mapr.com Kafka implementation 

I am not understanding. The only supported Kafka implementation is https://kafka.apache.org/

Maybe `mapr.com Kafka` belongs as a new input and output plugin?

**Comment by anelson-vidscale** [comment]
Thanks for looking Jordan!   :)

Presumably the existing Kafka plugin targets the API.
The MapR implementation implements that API.
I'm not asking you to test MapR's implementation of Kafka.
I'm just giving you the use case.

Adding support into logstash.yaml or logstash.jvm for extending the classpath seems both like a reasonable request and a natural way to solve this problem.  I might even venture to say that there are other use cases where one might wish a bit more control over the classpath than is currently offered by logstash.yaml.

FWIW, the "hack" of putting softlinks to the jars that I need to be included has been working just fine.  I never got the CLASSPATH approach to work.  It seems that one can replace the CLASSPATH, but not extend the CLASSPATH.  And clearly I need what is in vendor/jruby/lib to remain in CLASSPATH and anything else that logstash devs may decide needs to go into CLASSPATH.

**Comment by jordansissel** [comment]
> Presumably the existing Kafka plugin targets the API.

do you mean the kafka clients api? or the kafka broker api?

>  And clearly I need what is in vendor/jruby/lib to remain in CLASSPATH 

This is not a public interface we promise, so it may break (by design or by accident) in the future. I am glad you have a workaround for today, though.

**Comment by anelson-vidscale** [comment]
I mean presumably you target the kafka client API.  Presumably, you are not doing anything with the broker, but even if you are, it's still fine with the MapR implementation.

I know it's not a public interface that you promise.  That's why I'd like the CLASSPATH extension RFE.  That would be a "promised public interface".

**Comment by ryanjok** [comment]
@jordansissel Alex at Vidscale is requesting an update on this discussion to see where we are and where we stand as of today.

I would like to know if there is now an ElasticSearch sanctioned way to tweak CLASSPATH for logstash?
As we try to upgrade to Kafka 1.0 as part of upgrading our MapR implementation, we are running into problems again.

I just want to be able to tell the classpath in LogStash Java application that it can find jars in a few more places.
The fact that I'm substituting my jars (mapr jars) for the apache jars that implement the interface targeted by the kafka output plugin is just an example.

Having to write a new kafka output plugin just to control the classpath is silly.

Please reconsider your position.

That said, if you want me to repeat what folks are doing now , you can read this page: [https://community.mapr.com/docs/DOC-1468](https://community.mapr.com/docs/DOC-1468)

**Comment by jordansissel** [comment]
> an approach that feels hacky and un-blessed by elastic.co.

Modifying the CLASSPATH of a supported plugin (Kafka) by swapping out the implementation provided by the `org.apache.kafka` java namespace will make it an unsupported plugin.

I'm not against making it possible to inject your own classes, but I'm understanding how this helps you. The conversation so far seems to be customer needs in an unsupported area. Even if a different way is provided to swap out Logstash-provided java classes with your own (MapR), you're injecting unsupported code (MapR) into a product and then executing it. 

No matter how that injection is accomplished, MapR's Kafka client implementation is still unsupported.

We can decouple these two things:

1) Allow a way to inject arbitrary code into the classpath, as requested, with some to-be-determined interface.
2) Provide support for MapR's kafka client.

Even if we satisfy the first point, the second point (MapR) is still unsupported. Please let me know what you think about these points and your comfort level here.

@jsvd @acchen97 Your input may be useful here.

**Comment by anelson-vidscale** [comment]
I agree that 1. and 2 are separate.
I hope elastic.co will provide sanctioned methods for CLASSPATH manipulation which 1.
I agree that elastic.co is in no way required to support 2.

My hope is that if elastic.co adds support for 1, that will make it easy for VidScale and/or MapR to make forward progress.

**Comment by robfr** [comment]
What are the (unsactioned) methods to use CLASSPATH with logstash (6.6.0)? I have not succeded in setting it (via /etc/default/logstash for example).

In reading [the startup script](https://github.com/elastic/logstash/blob/0efc0d58d570522bd9a3cc7f28fb80e4650d3b8c/bin/logstash#L61) it seems as though any CLASSPATH environment variable is doomed to be ignored.

I ended up dropping my own jar into the logstash-core/lib/jars folder which gets the job done but is rather ugly.

**Comment by anelson-vidscale** [comment]
And we (edgegravity.ericsson.com) added softlinks to the external jars to the vendor/jruby/lib classpath.

It appears that making progress on  1. (support injecting other jars into CLASSPATH) did not happen.  :-(

Whenever we upgrade to a new version of Logstash, making sure that our vendor/jruby/ib hack still works is one of our first validation steps

**Comment by sydoracle** [comment]
An additional situation regarding this is accessing Wildfly (JBoss) JMX using the JMX input plugin
Wildfly requires an additional jboss-client.jar for a client to [access JMX](https://docs.wildfly.org/16/Admin_Guide.html#JMX) and putting the jar into vendor/jruby/lib is the [workaround](https://developer.jboss.org/thread/275310) 

This seems a similar case of trying to stretch an existing plugin to match an atypical endpoint.
